consumergroup.go

```diff
@@ -0,0 +1,249 @@
+}
+
+// IsExpired is a function to check if the lease is expired, but is only expected to be used in the heartbeat loop
+func (lease Lease) IsExpired(maxLeaseDuration time.Duration) bool {
+	if !lease.LastUpdateTime.IsZero() {
+		durationPast := time.Since(lease.LastUpdateTime)
+		if durationPast > maxLeaseDuration {
+			return true
+		}
+	}
+	return false
+}
+
+// CheckpointStorage is a simple interface for abstracting away the storage functions
+type CheckpointStorage interface {
+	CreateLease(lease Lease) error
+	UpdateLease(originalLease, updatedLease Lease) error
+	GetLease(leaseKey string) (*Lease, error)
+	GetAllLeases() (map[string]Lease, error)
+}
+
+// ConsumerGroupCheckpoint is a simple struct for managing the
+type ConsumerGroupCheckpoint struct {
+	Storage           CheckpointStorage
+	kinesis           Kinesis
+	LeaseDuration     time.Duration
+	HeartBeatDuration time.Duration
+	OwnerID           string
+	done              chan struct{}
+	currentLeases     map[string]*Lease //Initially, this will only be one
+	Mutex             *sync.Mutex
+}
+
+func (cgc ConsumerGroupCheckpoint) Get(shardID string) (string, error) {
+	return cgc.currentLeases[shardID].Checkpoint, nil
+}
+
+func (cgc ConsumerGroupCheckpoint) Set(shardID, sequenceNumber string) error {
+	cgc.Mutex.Lock()
+	defer cgc.Mutex.Unlock()
+
+	cgc.currentLeases[shardID].Checkpoint = sequenceNumber
+	return nil
+}
+
+func NewConsumerGroupCheckpoint(
+	storage CheckpointStorage,
+	kinesis Kinesis,
+	leaseDuration time.Duration,
+	heartBeatDuration time.Duration) *ConsumerGroupCheckpoint {
+	return &ConsumerGroupCheckpoint{
+		Storage:           storage,
+		kinesis:           kinesis,
+		LeaseDuration:     leaseDuration,
+		HeartBeatDuration: heartBeatDuration,
+		OwnerID:           uuid.NewV4().String(), // generated owner id
+		done:              make(chan struct{}),
+		currentLeases:     make(map[string]*Lease, 1),
+		Mutex:             &sync.Mutex{},
+	}
+
+}
+
+// Start is a blocking call that will attempt to acquire a lease on every tick of leaseDuration
+// If a lease is successfully acquired it will be added to the channel otherwise it will continue to retry
+func (cgc ConsumerGroupCheckpoint) Start(ctx context.Context, shardc chan string) {
+	fmt.Printf("Starting ConsumerGroupCheckpoint for Consumer %s \n", cgc.OwnerID)
+
+	tick := time.NewTicker(cgc.LeaseDuration)
+	defer tick.Stop()
+	var currentLeases map[string]Lease
+	var previousLeases map[string]Lease
+
+	for {
+		select {
+		case <-tick.C:
+			if len(cgc.currentLeases) == 0 { // only do anything if there are no current leases
+				fmt.Printf("Attempting to acquire lease for OwnerID=%s\n", cgc.OwnerID)
+				var err error
+				currentLeases, err = cgc.Storage.GetAllLeases()
+				if err != nil {
+					// TODO log this error
+				}
+
+				lease := cgc.CreateOrGetExpiredLease(currentLeases, previousLeases)
+				if lease != nil && lease.LeaseKey != "" {
```

Comment:
can we invert this conditional too? ideally the end of this loop will be the core functionality and not  indented

Reply:
Done